src/kubernetes.cc

```diff
@@ -793,33 +793,48 @@ void KubernetesReader::WatchMaster(
   request << boost::network::header(
       "Authorization", "Bearer " + KubernetesApiToken());
   const bool verbose = config_.VerboseLogging();
-  if (verbose) {
-    LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint;
-  }
-  try {
-    if (verbose) {
-      LOG(INFO) << "Locking completion mutex";
-    }
-    // A notification for watch completion.
-    std::mutex completion_mutex;
-    std::unique_lock<std::mutex> watch_completion(completion_mutex);
-    Watcher watcher(
-      endpoint,
-      [=](json::value raw_watch) {
-        WatchEventCallback(callback, name, std::move(raw_watch));
-      },
-      std::move(watch_completion), verbose);
-    http::client::response response = client.get(request, std::ref(watcher));
-    if (verbose) {
-      LOG(INFO) << "Waiting for completion";
-    }
-    std::lock_guard<std::mutex> await_completion(completion_mutex);
-    if (verbose) {
-      LOG(INFO) << "WatchMaster(" << name << ") completed " << body(response);
+  int failures = 0;
+  while (true) {
+    try {
+      if (verbose) {
+        LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint
+                  << " after " << failures << " failures";
+      }
+      // A notification for watch completion.
+      std::mutex completion_mutex;
+      std::unique_lock<std::mutex> watch_completion(completion_mutex);
+      Watcher watcher(
+        endpoint,
+        [=](json::value raw_watch) {
+          WatchEventCallback(callback, name, std::move(raw_watch));
+        },
+        std::move(watch_completion), verbose);
+      http::client::response response = client.get(request, std::ref(watcher));
+      if (verbose) {
+        LOG(INFO) << "Waiting for completion";
+      }
+      std::lock_guard<std::mutex> await_completion(completion_mutex);
+      if (verbose) {
+        LOG(INFO) << "WatchMaster(" << name << ") completed " << body(response);
+      }
+      // Connection closed without an error; reset failure count.
+      failures = 0;
+    } catch (const boost::system::system_error& e) {
+      LOG(ERROR) << "Failed to query " << endpoint << ": " << e.what();
+      ++failures;
+      if (failures >= config_.KubernetesUpdaterWatchConnectionRetries()) {
+        if (verbose) {
+          LOG(INFO) << "WatchMaster(" << name << "): Exiting after "
```

Comment:
Should this be an unconditional `LOG(ERROR)`?

Reply:
Done.